### PR TITLE
add Debian package configuration

### LINKED ISSUE #32: Support Debian packages
There are packages for debian in testing and backports; it would be very nice if they were used.

### PR DESCRIPTION
**Pull request by anarcat**
#### Pull Request (PR) description

This creates the necessary defaults on Debian to install Prometheus with the right settings from the Debian package.

The only major difference is the systemd unit file is kept in `/etc/systemd/system`: that is outside the scope of this change, as it's part of the `systemd` module.

#### This Pull Request (PR) fixes the following issues

Fixes: #32, #323 

#### Remaining work

I was able to install the node-exporter with this configuration using
standard Debian packages. This is a Work In Progress (WIP) because I
haven't tested the other packages and especially the main prometheus
package. Hopefully it shouldn't be too much work to implement that as
well.

The change is mostly a noop when running on top of node-exporter already installed by hand using normal Debian packages.

Tests and documentation might also need an update.

### CODE REVIEW

**Comment by bastelfreak** [comment]
@anarcat can you please take a look at the failing travis jobs?

**Comment by anarcat** [comment]
it's partly stuff that has to do with the path changes, for example:

>        expected that the catalogue would contain File[/usr/local/bin/node_exporter]

because it changes the defaults in the Debian environment, it affects basically the entire test suite. i'm not sure how to handle that...

**Comment by anarcat** [comment]
there's another thing i'm uncertain about: the Debian packages for the *-exporter stuff all use the same `prometheus` user: they don't create exporter-specific users the way the module does otherwise. this is what I tried to solve with:

```puppet
prometheus::node_exporter::group: 'prometheus'
prometheus::node_exporter::user: 'prometheus' 
```

... but I haven't done that with the apache exporter, for example. should we sync those up or does it matter?

similarly, the debian package already ship a `.service` file in `/lib` but the module installs a new one in `/etc` which shadows and overrides the one provided by the debian package. does that matter?

**Comment by bastelfreak** [comment]
service files that come from a package manager should go into `/lib`, custom services files should go into `/etc`. If the upstream service files work, we should probably use them and don't overwrite them.

**Comment by anarcat** [comment]
> service files that come from a package manager should go into /lib, custom services files should go into /etc. If the upstream service files work, we should probably use them and don't overwrite them.

so how should we do this? the default path is set in the `systemd` module and we call into `systemd::unit_file` from `prometheus::daemon`.

what you seem to be saying is that we should just not write a `.service` file at all if we're installing through a package is that correct? but then a bunch of stuff could break if (for example) the username is changed.

right now, the user override I did for the node exporter was to avoid creating a user, but the apache exporter would have been broken if we didn't install our own .service file...

**Comment by bastelfreak** [comment]
I think that a parameter like `manage_systemd_unit_file` might be a good idea. This should default to false, if `install_method` is `package`. But I'm not sure how easy it is to workaround the different user/groupnames.

**Comment by anarcat** [comment]
if i would have a clue of how to do unit tests, i would design a unit test that would:

 1. install everything through debian package (`*::install_method: package`)
 2. run the puppet manifests and make sure systemd units, users and groups are *not* created
 3. that everything still works:
    * that the processes get started with the right flags
    * that they run as the right user
    * etc

This is partly why this is WIP: I have no idea how to do this. I've been learning some spec chops in #304 but i'm nowhere near the level where I could do this...

**Comment by anarcat** [comment]
i have updated the branch to just bite the bullet and disable the .service file when install_method is through a package. no unit tests, but we should make sure things work somehow, i don't know... I also had to cleanup the old files using such a resource in the parent profile:

```puppet
  # cleanup old .service file previously deployed by prometheus module
  include systemd
  file {
    '/etc/systemd/system/prometheus.service':
      ensure => absent,
      notify => Class['systemd::systemctl::daemon_reload'],
  }
```

similar structures for the exporter, of course... that feels a little redundant - i wonder if we shouldn't enforce that in the class instead of wrapping the whole systemd thing inside an `if`. but then it would destroy local customizations people might have as well, so maybe it's better to let the users handle the transition since we don't support debian packages officially anyways right now...

File: manifests/config.pp [hunk 1]
```diff
@@ -67,9 +67,11 @@
         }
       }
       'systemd' : {
-        include 'systemd'
-        systemd::unit_file {'prometheus.service':
-          content => template('prometheus/prometheus.systemd.erb'),
+        if $prometheus::server::manage_systemd_unit {
+          include 'systemd'
+          systemd::unit_file {'prometheus.service':
```
Comment (bastelfreak) [inline manifests/config.pp]: instead of defining the `systemd::unit_file` resoure here, should we only set `$ensure = 'present` and pass that to the resource? And if `$prometheus::server::manage_systemd_unit` isn't true, set `$ensure = 'absent'`.

Comment (anarcat) [reply manifests/config.pp]: that's what i said earlier:

> i wonder if we shouldn't enforce that in the class instead of wrapping the whole systemd thing inside an `if`. but then it would destroy local customizations people might have as well, so maybe it's better to let the users handle the transition since we don't support debian packages officially anyways right now...

i don't mind either way...

**Comment by anarcat** [comment]
what should be my next step here?

**Comment by anarcat** [comment]
> I think that a parameter like manage_systemd_unit_file might be a good idea. This should default to false, if install_method is package. But I'm not sure how easy it is to workaround the different user/groupnames.

Another problem has come up with this in #310: the way parameters are passed in the Debian service file is through the `$ARGS` variable set in the `EnvironmentFile` (`/etc/default/prometheus-postfix-exporter`). The `.service` file deployed by the puppet modules doesn't use that mechanism to pass flags which makes it difficult to have a coherent approach for both.

For now I've used the `env_vars` parameter for the postfix service, but it would be nice to have a more uniform way to pass `options` around between the two ways of deploying exporters.

**Comment by dhoppe** [comment]
@anarcat Could you please do a rebase?

**Comment by anarcat** [comment]
@dhoppe done, but i'd still need help and guidance on the next steps to move forward here. there are unresolved conversations and questions above I would love to have feedback on. :)

File: data/Debian.yaml [hunk 2]
```diff
@@ -1,2 +1,19 @@
 ---
 prometheus::env_file_path: '/etc/default'
+prometheus::bin_dir: '/usr/bin'
+prometheus::shared_dir: '/usr/share/prometheus'
+prometheus::install_method: 'package'
```
Comment (dhoppe) [inline data/Debian.yaml]: @anarcat Why do change the install method for Debian? It makes this pull request backwards incompatible and can be set by yourself, if necessary.

File: data/defaults.yaml [hunk 3]
```diff
@@ -57,7 +57,7 @@ prometheus::alertmanager::user: 'alertmanager'
 prometheus::alertmanager::version: '0.5.1'
 prometheus::alerts: {}
 prometheus::config_dir: '/etc/prometheus'
-prometheus::config_mode: '0660'
+prometheus::config_mode: '0640'
```
Comment (dhoppe) [inline data/defaults.yaml]: @anarcat If you change the permissions, you need to modify the file ```spec/classes/prometheus_spec.rb``` aswell.

File: manifests/config.pp [hunk 4]
```diff
@@ -131,7 +133,6 @@
     file { 'prometheus.yaml':
       ensure       => present,
       path         => "${prometheus::server::config_dir}/${prometheus::server::configname}",
-      owner        => $prometheus::server::user,
```
Comment (dhoppe) [inline manifests/config.pp]: @anarcat Why did you remove the owner? I am pretty sure this is set on purpose.

File: manifests/server.pp [hunk 5]
```diff
@@ -62,7 +63,6 @@
 
   file { "${config_dir}/rules":
     ensure => 'directory',
-    owner  => $user,
```
Comment (dhoppe) [inline manifests/server.pp]: @anarcat Why did you remove the owner? I am pretty sure this is set on purpose.

File: data/Debian.yaml (hunk 2 above)
Comment (anarcat) [reply data/Debian.yaml]: i'm trying to reduce the diff with my local stuff. but if that would help get this through, sure, i can revert this...

File: manifests/config.pp (hunk 4 above)
Comment (anarcat) [reply manifests/config.pp]: because i don't want the file to be writable by the prometheus user. services shouldn't be able to modify their own config files.

File: manifests/server.pp (hunk 5 above)
Comment (anarcat) [reply manifests/server.pp]: same answer.

File: data/defaults.yaml (hunk 3 above)
Comment (anarcat) [reply data/defaults.yaml]: thanks for the pointer! i've updated the test suite, which should hopefully reduce the number of failures :)

File: data/Debian.yaml (hunk 2 above)
Comment (anarcat) [reply data/Debian.yaml]: so i tried reverting this part, but it also involved removing a bunch more configuration, as you can see in 2e70883 - it doesn't feel right...

File: data/defaults.yaml (hunk 3 above)
Comment (anarcat) [reply data/defaults.yaml]: i split this question off into a separate, more elaborate PR in #324

File: manifests/config.pp (hunk 4 above)
Comment (anarcat) [reply manifests/config.pp]: i split this question off into a separate, more elaborate PR in #324

**Comment by anarcat** [comment]
I have greatly simplified this by ripping out the permissions tweak in a separate PR (#323) since that is probably relevant to all installs, not just Debian.

I have also removed the Hiera customization that is associated with the Debian package setup. That is trickier, because it's hard to get right for users. To enable the Debian package, they would need to do this chunk:

```
prometheus::bin_dir: '/usr/bin'
prometheus::shared_dir: '/usr/share/prometheus'
prometheus::install_method: 'package'
prometheus::node_exporter::package_ensure: 'latest'
prometheus::node_exporter::package_name: 'prometheus-node-exporter'
prometheus::node_exporter::service_name: 'prometheus-node-exporter'
prometheus::node_exporter::group: 'prometheus'
prometheus::node_exporter::user: 'prometheus'
prometheus::apache_exporter::package_name: 'prometheus-apache-exporter'
prometheus::apache_exporter::service_name: 'prometheus-apache-exporter'
prometheus::apache_exporter::user: 'prometheus'
prometheus::apache_exporter::group: 'prometheus'
prometheus::configname: 'prometheus.yml'
prometheus::server::user: 'root'
prometheus::server::group: 'root'
prometheus::postgres_exporter::package_name: 'prometheus-postgres-exporter'
prometheus::postgres_exporter::service_name: 'prometheus-postgres-exporter'
prometheus::postgres_exporter::user: 'prometheus'
prometheus::postgres_exporter::group: 'prometheus'
```

... I think you'll agree this is preeeetty annoying and not very practical. I am not sure how best to do this cleanly without flipping the default install method on Debian, to be honest, since we can't have conditionals in YAML (or can we?).

Would love to have a better way to do this, but failing that, this will have to do for us for now, because i'd really like to see at least *some* of this get through.

**Comment by anarcat** [comment]
tests should now be passing here, whoohoo! i'll remove the WIP tag so we can start taking this one more seriously.

**Comment by anarcat** [comment]
tests pass here, can we remove the test-fail label and get a final review?

File: manifests/daemon.pp [hunk 6]
```diff
@@ -82,6 +82,7 @@
   String $service_ensure               = 'running',
   Boolean $service_enable              = true,
   Boolean $manage_service              = true,
+  Boolean $manage_systemd_unit         = $install_method != 'package',
```
Comment (ekohl) [inline manifests/daemon.pp]: I never know if this actually works or if it's trying to be too clever. Usually I play it safe, set it to `undef` and use `pick($manage_systemd_unit, $install_method != 'package')`

File: templates/prometheus.systemd.erb [hunk 7]
```diff
@@ -7,8 +7,13 @@ After=basic.target network.target
 [Service]
 User=<%= scope.lookupvar('prometheus::server::user') %>
 Group=<%= scope.lookupvar('prometheus::server::group') %>
+<% if scope.lookupvar('prometheus::server::env_file_path') -%>
+EnvironmentFile=<%= scope.lookupvar('prometheus::server::env_file_path') %>/prometheus
```
Comment (ekohl) [inline templates/prometheus.systemd.erb]: Does the file need to exist? Otherwise you can use
```suggestion
EnvironmentFile=-<%= scope.lookupvar('prometheus::server::env_file_path') %>/prometheus
```

Also note that systemd deprecated `EnvironmentFile` in favor of drop in files. Then you can use `Environment=ARGS=...` in this file and it can be overridden in a drop in.

File: manifests/daemon.pp (hunk 6 above)
Comment (anarcat) [reply manifests/daemon.pp]: not being familiar with the `pick` function, I don't actually know how you would expect me to use this. :)

File: templates/prometheus.systemd.erb (hunk 7 above)
Comment (anarcat) [reply templates/prometheus.systemd.erb]: > Does the file need to exist?

... i think so? but it's better to be safe than sorry i guess... we *do* create the file elsewhere, if my memory is correct, however... so maybe it's better to actually fail if we fail to create it?

> Also note that systemd deprecated EnvironmentFile in favor of drop in files. Then you can use Environment=ARGS=... in this file and it can be overridden in a drop in.

Noted. The point here, however, is to have a single env file that works across init.d and systemd configurations...

File: manifests/daemon.pp (hunk 6 above)
Comment (ekohl) [reply manifests/daemon.pp]: https://github.com/puppetlabs/puppetlabs-stdlib/blob/master/REFERENCE.md#pick

My biggest concern is that your implementation relies on a certain compile order and I don't know if Puppet actually guarantees that it works. If you override install_method via hiera, does it also still work?

**Comment by bastelfreak** [comment]
Hey people. I'm not really sure how we should proceed here.

**Comment by anarcat** [comment]
i've spun off the apache exporter service_name tweak to a separate PR, in #400, because I think that's a reasonable fix to do, regardless of how you feel about the more controversial parts of this PR.

Since @alexjfisher implemented the `init_style` parameter in #399, all that is left here is to deploy and use the `env_vars` file (e.g. `/etc/default/prometheus`) in Debian. I am not sure I understand correctly how that file is supposed to work, but this works for me.

By the way, I do understand it might be better to override the systemd unit in `/etc/systemd/system/overrides.d`, but that's not what were doing now: we're *shadowing* the unit file from `/lib` and I think that's the wrong way to go, because we're re-shipping an existing configuration file that might break when deployed with the debian package.

Maybe the right way to go would be to deploy a overrides file instead of shadowing it? I'd be okay with such a compromise, for what that's worth...

It's been a long time I started on this project (9 months!) so I don't exactly remember what broke when I was trying to deploy this with the debian package. But I do remember it broke, so I think we do need to use something like this to make things work correctly.

And I must say I have been using this code for those 9 months in production, and I'm pretty happy with it. Obviously, the tests still fail so it's likely that this introduces problems in other platforms (or maybe just in the unit tests themselves), but i thought that was worth mentioning...

**Comment by anarcat** [comment]
i discussed this again with @alexjfisher last week, and it seems the three options are:

 1. accept that we shadow `/lib` systemd files with our own, but at least make sure they are in sync
 2. allow using the `/lib` systemd files and just use *overrides* in `/etc/systemd/system/overrides.d` to deploy the various parameters configurable in `prometheus::daemon` and friends, and only if `install_method=='package'`?
 3. keep plowing ahead with the current approach of (ab?)using the `env_vars` file

i would love to hear from others which direction you think we should take here.

**Comment by anarcat** [comment]
it looks like we agreed on option 3 yesterday. the way forward should probably imply starting from scratch here, and accomplish the following:

 1. [ ] `install_method==package` should imply `init_style=none`?
 2. [ ] deploy an `env_vars` file in `prometheus::server` so that the Debian package unit file can get the arguments through (probably if `install_method==package`)
 3. [ ] deploy the `env_vars` file in the postfix exporter and pass it the `extra_options` list
 4. [ ] possibly, pass the `$options` to a `ARGS` entry in `env_vars` if `init_style=none`?

Then tests will probably fail and need some meddling...

**Comment by bastelfreak** [comment]
I am not sure if `install_method='package'` should imply `init_style='none'`. On Arch we install the packages, but use the unit file from this module. our daemon.pp supports an env_file. We need to ensure that every module has the `$env_vars` and `$env_file_path` variable. I am also not sure about passing `$args` to the environment file, because for that purpose we already have the `$env_vars` variable. But all in all, this should work fine.

**Comment by anarcat** [comment]
i just noticed that env_vars isn't passed to the blackbox exporter as well now, which makes it fail to start completely here. so we'd probably need to add it *all* over the place :/

**Comment by bastelfreak** [comment]
this fixes most of the tests:

```patch
diff --git a/templates/prometheus.systemd.epp b/templates/prometheus.systemd.epp
index 83e2d12..1728940 100644
--- a/templates/prometheus.systemd.epp
+++ b/templates/prometheus.systemd.epp
@@ -14,13 +14,13 @@ After=basic.target network.target
 [Service]
 User=<%= $user %>
 Group=<%= $group %>
-<% if $env_file_path -%>
+<% if $env_file_path {-%>
 EnvironmentFile=<%= $env_file_path %>/prometheus
 ExecStart=<%= $bin_dir %>/prometheus $ARGS
-<% else -%>
+<% } else { -%>
 ExecStart=<%= $bin_dir %>/prometheus \
   <%= $daemon_flags.join(" \\\n  ") %>
-<% end -%>
+<% } -%>
 ExecReload=/bin/kill -HUP $MAINPID
 KillMode=process
 Restart=always
```

**Comment by bastelfreak** [comment]
and some more:
```
diff --git a/manifests/config.pp b/manifests/config.pp
index 8129672..bbbbe1f 100644
--- a/manifests/config.pp
+++ b/manifests/config.pp
@@ -164,7 +164,7 @@ class prometheus::config {
       mode    => '0644',
       owner   => 'root',
       group   => '0', # Darwin uses wheel
-      content => "ARGS='${join($daemon_flags, ' ')}'\n",
+      content => "ARGS='${join(sort($daemon_flags), ' ')}'\n",
     }
   }
 
@@ -192,7 +192,7 @@ class prometheus::config {
         content => epp("${module_name}/prometheus.systemd.epp", {
           'user'           => $prometheus::server::user,
           'group'          => $prometheus::server::group,
-          'daemon_flags'   => $daemon_flags,
+          'daemon_flags'   => $daemon_flags.sort,
           'max_open_files' => $max_open_files,
           'bin_dir'        => $prometheus::server::bin_dir,
           'env_file_path'  => $prometheus::server::env_file_path,
diff --git a/spec/fixtures/files/prometheus.default b/spec/fixtures/files/prometheus.default
index c2f631e..e68d2f6 100644
--- a/spec/fixtures/files/prometheus.default
+++ b/spec/fixtures/files/prometheus.default
@@ -1 +1 @@
-ARGS='--config.file=/etc/prometheus/prometheus.yaml --storage.tsdb.path=/var/lib/prometheus --storage.tsdb.retention=360h --web.console.templates=/usr/local/share/prometheus/consoles --web.console.libraries=/usr/local/share/prometheus/console_libraries'
+ARGS='--config.file=/etc/prometheus/prometheus.yaml --storage.tsdb.path=/var/lib/prometheus --storage.tsdb.retention=360h --web.console.libraries=/usr/local/share/prometheus/console_libraries --web.console.templates=/usr/local/share/prometheus/consoles'
```

File: manifests/config.pp [hunk 8]
```diff
@@ -159,6 +159,15 @@
     default => undef,
   }
 
+  if $prometheus::env_file_path != undef {
```
Comment (bastelfreak) [inline manifests/config.pp]: we can shorten this:

```suggestion
  if $prometheus::env_file_path {
```

Comment (anarcat) [reply manifests/config.pp]: done, thanks

**Comment by anarcat** [comment]
> this fixes most of the tests:

committed.

> and some more:

committed.

**Comment by anarcat** [comment]
alright, i split this up in two smaller PRs: #527 and #528. Neither touches the test suite, which means the tests will probably pass, but that's kind of a little white lie: the code they affect (install_method=package and init_style=none) is basically untested anyways.

but at least it should be mergeable, and should form a better basis to fix #32 in the long run than this too old and complicated PR. i'll regroup on those and summarize progress in #32 instead of keeping this old thing alive.
